treeherder/etl/taskcluster_pulse/handler.py

```diff
@@ -0,0 +1,295 @@
+    }
+
+    if pushInfo["origin"] == "hg.mozilla.org":
+        job["origin"]["pushLogID"] = int(pushInfo["pushId"])
+    else:
+        job["origin"]["pullRequestID"] = pushInfo["pushId"]
+        job["origin"]["owner"] = pushInfo["owner"]
+
+    # Transform "collection" into an array of labels if task doesn't
+    # define "labels".
+    labels = treeherderConfig.get("labels", [])
+    if not labels:
+        if not treeherderConfig.get("collection"):
+            labels = ["opt"]
+        else:
+            labels = list(treeherderConfig["collection"].keys())
+
+    job["labels"] = labels
+
+    machine = treeherderConfig.get("machine", {})
+    job["buildMachine"] = {
+      "name": jobRun.get("workerId", "unknown"),
+      "platform": machine.get("platform", task["workerType"]),
+      "os": machine.get("os", "-"),
+      "architecture": machine.get("architecture", "-"),
+    }
+
+    if treeherderConfig.get("productName"):
+        job["productName"] = treeherderConfig["productName"]
+
+    if treeherderConfig.get("groupName"):
+        job["display"]["groupName"] = treeherderConfig["groupName"]
+
+    return job
+
+
+def handleTaskPending(pushInfo, task, message):
+    return buildMessage(pushInfo, task, message["runId"], message)
+
+
+def handleTaskRerun(pushInfo, task, message):
+    job = buildMessage(pushInfo, task, message["runId"]-1, message)
+    job["state"] = "completed"
+    job["result"] = "fail"
+    job["isRetried"] = True
+    # reruns often have no logs, so in the interest of not linking to a 404'ing artifact,
+    # don't include a link
+    job["logs"] = []
+    job = addArtifactUploadedLinks(
+        message["status"]["taskId"],
+        message["runId"]-1,
+        job)
+    return job
+
+
+def handleTaskRunning(pushInfo, task, message):
+    job = buildMessage(pushInfo, task, message["runId"], message)
+    job["timeStarted"] = message["status"]["runs"][message["runId"]]["started"]
+    return job
+
+
+def handleTaskCompleted(pushInfo, task, message):
+    jobRun = message["status"]["runs"][message["runId"]]
+    job = buildMessage(pushInfo, task, message["runId"], message)
+
+    job["timeStarted"] = jobRun["started"]
+    job["timeCompleted"] = jobRun["resolved"]
+    job["logs"] = [createLogReference(message["status"]["taskId"], jobRun["runId"])]
+    job = addArtifactUploadedLinks(
+        message["status"]["taskId"],
+        message["runId"],
+        job)
+    return job
+
+
+def handleTaskFailed(pushInfo, task, message):
+    jobRun = message["status"]["runs"][message["runId"]]
+    job = buildMessage(pushInfo, task, message["runId"], message)
+    job["timeStarted"] = jobRun["started"]
+    job["timeCompleted"] = jobRun["resolved"]
+    job["logs"] = [createLogReference(message["status"]["taskId"], jobRun["runId"])]
+    job = addArtifactUploadedLinks(
+        message["status"]["taskId"],
+        message["runId"],
+        job)
+    return job
+
+
+def handleTaskException(pushInfo, task, message):
```

Comment:
This is also super similar to ``handleTaskCompleted``.  Perhaps we could merge these three functions to handle the cases?

Reply:
I tried but it was not as straight forward.

I can handle this on a clean up in https://bugzilla.mozilla.org/show_bug.cgi?id=1560596